### PR TITLE
Set `dataModelProvider` as a required argument for `Controller::FactoryInitParams`

### PR DESCRIPTION
**Pull request by andy31415**
Similar to what #36558 did for server, this is the controller side.

### CODE REVIEW

File: src/controller/CHIPDeviceControllerFactory.cpp [hunk 1]
```diff
@@ -295,8 +304,11 @@ CHIP_ERROR DeviceControllerFactory::InitSystemState(FactoryInitParams params)
     stateParams.caseSessionManager = Platform::New<CASESessionManager>();
     ReturnErrorOnFailure(stateParams.caseSessionManager->Init(stateParams.systemLayer, sessionManagerConfig));
 
-    ReturnErrorOnFailure(chip::app::InteractionModelEngine::GetInstance()->Init(
-        stateParams.exchangeMgr, stateParams.fabricTable, stateParams.reportScheduler, stateParams.caseSessionManager));
+    chip::app::InteractionModelEngine * interactionModelEngine = chip::app::InteractionModelEngine::GetInstance();
+
+    interactionModelEngine->SetDataModelProvider(params.dataModelProvider);
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceControllerFactory.cpp:309]: This looks wrong.  InitDataModelHandler() happened above, and our ordering guarantee is that SetDataModelProvider() happens before that, no?  We should match the Server ordering of these things for now (and make sure to update the orders in sync).

That said, it's really weird that things are using the interaction model engine (to get the data model provider) before we have inited it.  We really need to think about the right way to keep track of these sorts of "global stack state".  But followup for that, obviously.

File: examples/chip-tool/commands/common/CHIPCommand.cpp [hunk 2]
```diff
@@ -137,6 +138,7 @@ CHIP_ERROR CHIPCommand::MaybeSetUpStack()
     factoryInitParams.opCertStore              = &mOpCertStore;
     factoryInitParams.enableServerInteractions = NeedsOperationalAdvertising();
     factoryInitParams.sessionKeystore          = &sSessionKeystore;
+    factoryInitParams.dataModelProvider        = chip::app::CodegenDataModelProviderInstance();
```
[the remark this answers is not in the document]
Comment (andy31415) [reply examples/chip-tool/commands/common/CHIPCommand.cpp:141]: Most will not, so I would like to eventually have a `Null/Empty` provider along with also building up to an aggregator that we can use. For now I biased towards `keep it like it was before`

File: src/controller/CHIPDeviceControllerFactory.cpp (hunk 1 above)
Comment (andy31415) [reply src/controller/CHIPDeviceControllerFactory.cpp:309]: Moved it higher above. I was relying on integration and unit tests to catch things and those passed even with a null data model (still fixing some unit tests).

File: src/controller/CHIPDeviceControllerFactory.cpp [hunk 3]
```diff
@@ -238,6 +250,9 @@ CHIP_ERROR DeviceControllerFactory::InitSystemState(FactoryInitParams params)
     ReturnErrorOnFailure(stateParams.unsolicitedStatusHandler->Init(stateParams.exchangeMgr));
     ReturnErrorOnFailure(stateParams.bdxTransferServer->Init(stateParams.systemLayer, stateParams.exchangeMgr));
 
+    chip::app::InteractionModelEngine * interactionModelEngine = chip::app::InteractionModelEngine::GetInstance();
+    interactionModelEngine->SetDataModelProvider(params.dataModelProvider);
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceControllerFactory.cpp:254]: Please document that this must come before InitDataModelHandler at the moment and why.

**Review by bzbarsky-apple — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

**Review by tcarmelveilleux — APPROVED** [review]

**Pull request merged**